File: src/barcode/code39.ts
import type { Color } from '../color/color.js';
import { renderBars, type Bar } from './barcode-renderer.js';

/**
 * Code 39 character encoding patterns.
 * Each character is encoded as 9 elements: 5 bars and 4 spaces.
 * 'n' = narrow, 'w' = wide. Pattern is: B S B S B S B S B
 * Characters: 0-9, A-Z, -, ., space, $, /, +, %, *
 */
const CHAR_PATTERNS: Record<string, string> = {
  '0': 'nnnwwnwnn',
  '1': 'wnnwnnnnw',
  '2': 'nnwwnnnnw',
  '3': 'wnwwnnnn' + 'n', // wnwwnnnnn
  '4': 'nnnwwnnnw',  // 'nnnwwnnnw' is incorrect, re-check
  '5': 'wnnwwnnnn',
  '6': 'nnwwwnnnn',
  '7': 'nnnwnnwnw',
  '8': 'wnnwnnwnn',
  '9': 'nnwwnnwnn',
  'A': 'wnnnnwnnw',
  'B': 'nnwnnwnnw',
  'C': 'wnwnnwnnn',
  'D': 'nnnnwwnnw',
  'E': 'wnnnwwnnn',
  'F': 'nnwnwwnnn',
  'G': 'nnnnnwwnw',
  'H': 'wnnnnwwnn',
  'I': 'nnwnnwwnn',
  'J': 'nnnnwwwnn',
  'K': 'wnnnnnnww',
  'L': 'nnwnnnnww',
  'M': 'wnwnnnnwn',
  'N': 'nnnnwnnww',
  'O': 'wnnnwnnwn',
  'P': 'nnwnwnnwn',
  'Q': 'nnnnnnwww',
  'R': 'wnnnnnwwn',
  'S': 'nnwnnnwwn',
  'T': 'nnnnwnwwn',
  'U': 'wwnnnnnnw',
  'V': 'nwwnnnnnw',
  'W': 'wwwnnnnnn',
  'X': 'nwnnwnnnw',
  'Y': 'wwnnwnnnn',
  'Z': 'nwwnwnnnn',
  '-': 'nwnnnnwnw',
  '.': 'wwnnnnwnn',
  ' ': 'nwwnnnwnn',
  '$': 'nwnwnwnnn',
  '/': 'nwnwnnnwn',
  '+': 'nwnnnwnwn',
  '%': 'nnnwnwnwn',
  '*': 'nwnnwnwnn',
};

// Corrected Code 39 patterns based on the standard
// Each character = 9 elements (BSBSBSBSB), 3 wide elements, 6 narrow
const CODE39_PATTERNS: Record<string, number[]> = {};

function initPatterns(): void {
  // Standard Code 39 encoding
  // Format: array of 9 widths (bar, space, bar, space, bar, space, bar, space, bar)
  // 1 = narrow, 2 = wide
  const defs: [string, number[]][] = [
    ['0', [1,1,1,2,2,1,2,1,1]],
    ['1', [2,1,1,2,1,1,1,1,2]],
    ['2', [1,1,2,2,1,1,1,1,2]],
    ['3', [2,1,2,2,1,1,1,1,1]],
    ['4', [1,1,1,2,2,1,1,1,2]],
    ['5', [2,1,1,2,2,1,1,1,1]],
    ['6', [1,1,2,2,2,1,1,1,1]],
    ['7', [1,1,1,2,1,1,2,1,2]],
    ['8', [2,1,1,2,1,1,2,1,1]],
    ['9', [1,1,2,2,1,1,2,1,1]],
    ['A', [2,1,1,1,1,2,1,1,2]],
    ['B', [1,1,2,1,1,2,1,1,2]],
    ['C', [2,1,2,1,1,2,1,1,1]],
    ['D', [1,1,1,1,2,2,1,1,2]],
    ['E', [2,1,1,1,2,2,1,1,1]],
    ['F', [1,1,2,1,2,2,1,1,1]],
    ['G', [1,1,1,1,1,2,2,1,2]],
    ['H', [2,1,1,1,1,2,2,1,1]],
    ['I', [1,1,2,1,1,2,2,1,1]],
    ['J', [1,1,1,1,2,2,2,1,1]],
    ['K', [2,1,1,1,1,1,1,2,2]],
    ['L', [1,1,2,1,1,1,1,2,2]],
    ['M', [2,1,2,1,1,1,1,2,1]],
    ['N', [1,1,1,1,2,1,1,2,2]],
    ['O', [2,1,1,1,2,1,1,2,1]],
    ['P', [1,1,2,1,2,1,1,2,1]],
    ['Q', [1,1,1,1,1,1,2,2,2]],
    ['R', [2,1,1,1,1,1,2,2,1]],
    ['S', [1,1,2,1,1,1,2,2,1]],
    ['T', [1,1,1,1,2,1,2,2,1]],
    ['U', [2,2,1,1,1,1,1,1,2]],
    ['V', [1,2,2,1,1,1,1,1,2]],
    ['W', [2,2,2,1,1,1,1,1,1]],
    ['X', [1,2,1,1,2,1,1,1,2]],
    ['Y', [2,2,1,1,2,1,1,1,1]],
    ['Z', [1,2,2,1,2,1,1,1,1]],
    ['-', [1,2,1,1,1,1,2,1,2]],
    ['.', [2,2,1,1,1,1,2,1,1]],
    [' ', [1,2,2,1,1,1,2,1,1]],
    ['$', [1,2,1,2,1,2,1,1,1]],
    ['/', [1,2,1,2,1,1,1,2,1]],
    ['+', [1,2,1,1,1,2,1,2,1]],
    ['%', [1,1,1,2,1,2,1,2,1]],
    ['*', [1,2,1,1,2,1,2,1,1]],
  ];

  for (const [ch, pattern] of defs) {
    CODE39_PATTERNS[ch] = pattern;
  }
}

initPatterns();

/**
 * Encode data as Code 39 barcode.
 * Automatically wraps data in start/stop '*' characters.
 * Returns boolean array: true = bar (dark module), false = space (light module).
 */
export function encodeCode39(data: string): boolean[] {
  const upper = data.toUpperCase();

  // Validate characters
  for (const ch of upper) {
    if (!CODE39_PATTERNS[ch]) {
      throw new Error(`Code 39: invalid character '${ch}'`);
    }
  }

  // Full string with start/stop
  const fullData = '*' + upper + '*';
  const result: boolean[] = [];

  for (let c = 0; c < fullData.length; c++) {
    if (c > 0) {
      // Inter-character gap (narrow space)
      result.push(false);
    }

    const pattern = CODE39_PATTERNS[fullData[c]];
    for (let i = 0; i < pattern.length; i++) {
      const width = pattern[i];
      const isBar = i % 2 === 0; // Even indices are bars, odd are spaces
      for (let w = 0; w < width; w++) {
        result.push(isBar);
      }
    }
  }

  return result;
}

/**
 * Encode data and render as PDF drawing operators.
 */
export function renderCode39(
  data: string,
  x: number,
  y: number,
  width: number,
  height: number,
  color?: Color,
): string {
  const bits = encodeCode39(data);
  const totalModules = bits.length;
  const moduleWidth = width / totalModules;

  const bars: Bar[] = [];
  for (let i = 0; i < bits.length; i++) {
    if (bits[i]) {
      bars.push({
        x: i * moduleWidth,
        width: moduleWidth,
        height,
      });
    }
  }

  return renderBars(bars, x, y, height, color);
}

File: src/barcode/code39.zig
const std = @import("std");
const Allocator = std.mem.Allocator;

/// Code 39 character patterns.
/// Each character is encoded as 9 elements (bars and spaces), where
/// a wide element = 3 and a narrow element = 1.
/// Format: bar, space, bar, space, bar, space, bar, space, bar
const CharPattern = [9]u8;

const char_set = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ-. $/+%*";

const char_patterns = [44]CharPattern{
    .{ 1, 1, 1, 3, 3, 1, 3, 1, 1 }, // 0
    .{ 3, 1, 1, 1, 1, 3, 1, 1, 3 }, // 1
    .{ 1, 1, 3, 1, 1, 3, 1, 1, 3 }, // 2
    .{ 3, 1, 3, 1, 1, 3, 1, 1, 1 }, // 3
    .{ 1, 1, 1, 1, 3, 3, 1, 1, 3 }, // 4
    .{ 3, 1, 1, 1, 3, 3, 1, 1, 1 }, // 5
    .{ 1, 1, 3, 1, 3, 3, 1, 1, 1 }, // 6
    .{ 1, 1, 1, 1, 1, 3, 3, 1, 3 }, // 7
    .{ 3, 1, 1, 1, 1, 3, 3, 1, 1 }, // 8 (value=8, char='8')
    .{ 1, 1, 3, 1, 1, 3, 3, 1, 1 }, // 9
    .{ 3, 1, 1, 3, 1, 1, 1, 1, 3 }, // A
    .{ 1, 1, 3, 3, 1, 1, 1, 1, 3 }, // B
    .{ 3, 1, 3, 3, 1, 1, 1, 1, 1 }, // C
    .{ 1, 1, 1, 3, 3, 1, 1, 1, 3 }, // D
    .{ 3, 1, 1, 3, 3, 1, 1, 1, 1 }, // E
    .{ 1, 1, 3, 3, 3, 1, 1, 1, 1 }, // F
    .{ 1, 1, 1, 1, 1, 3, 1, 3, 3 }, // G (value=16)
    .{ 3, 1, 1, 1, 1, 1, 1, 3, 3 }, // H (value=17) - not used in indexing below
    .{ 1, 1, 3, 1, 1, 1, 1, 3, 3 }, // I
    .{ 1, 1, 1, 1, 3, 1, 1, 3, 3 }, // J
    .{ 3, 1, 1, 1, 1, 1, 3, 1, 3 }, // K (value=20)
    .{ 1, 1, 3, 1, 1, 1, 3, 1, 3 }, // L
    .{ 3, 1, 3, 1, 1, 1, 3, 1, 1 }, // M
    .{ 1, 1, 1, 1, 3, 1, 3, 1, 3 }, // N
    .{ 3, 1, 1, 1, 3, 1, 3, 1, 1 }, // O
    .{ 1, 1, 3, 1, 3, 1, 3, 1, 1 }, // P
    .{ 1, 1, 1, 1, 1, 1, 3, 3, 3 }, // Q
    .{ 3, 1, 1, 1, 1, 1, 3, 3, 1 }, // R
    .{ 1, 1, 3, 1, 1, 1, 3, 3, 1 }, // S
    .{ 1, 1, 1, 1, 3, 1, 3, 3, 1 }, // T
    .{ 3, 3, 1, 1, 1, 1, 1, 1, 3 }, // U (value=30)
    .{ 1, 3, 3, 1, 1, 1, 1, 1, 3 }, // V
    .{ 3, 3, 3, 1, 1, 1, 1, 1, 1 }, // W
    .{ 1, 3, 1, 1, 3, 1, 1, 1, 3 }, // X
    .{ 3, 3, 1, 1, 3, 1, 1, 1, 1 }, // Y
    .{ 1, 3, 3, 1, 3, 1, 1, 1, 1 }, // Z
    .{ 1, 3, 1, 1, 1, 1, 3, 1, 3 }, // -
    .{ 3, 3, 1, 1, 1, 1, 3, 1, 1 }, // .
    .{ 1, 3, 1, 1, 1, 1, 1, 3, 3 }, // (space) (value=38)
    .{ 1, 3, 1, 3, 1, 3, 1, 1, 1 }, // $ (value=39)
    .{ 1, 3, 1, 3, 1, 1, 1, 3, 1 }, // / (value=40)
    .{ 1, 3, 1, 1, 1, 3, 1, 3, 1 }, // + (value=41)
    .{ 1, 1, 1, 3, 1, 3, 1, 3, 1 }, // % (value=42)
    .{ 1, 3, 1, 1, 3, 1, 3, 1, 1 }, // * (start/stop) (value=43)
};

const start_stop_idx: usize = 43; // '*' character

/// Look up the pattern index for a given character.
fn charIndex(ch: u8) ?usize {
    for (char_set, 0..) |c, i| {
        if (c == ch) return i;
    }
    return null;
}

/// Encode a string value into Code 39 bar widths.
/// The value should contain only valid Code 39 characters (0-9, A-Z, -, ., space, $, /, +, %).
/// Start/stop characters (*) are added automatically.
pub fn encode(allocator: Allocator, value: []const u8) ![]u8 {
    var result: std.ArrayListUnmanaged(u8) = .{};
    errdefer result.deinit(allocator);

    // Start character
    try result.appendSlice(allocator, &char_patterns[start_stop_idx]);
    try result.append(allocator, 1); // Inter-character gap (narrow space)

    for (value) |ch| {
        const upper = if (ch >= 'a' and ch <= 'z') ch - 32 else ch;
        const idx = charIndex(upper) orelse continue;
        try result.appendSlice(allocator, &char_patterns[idx]);
        try result.append(allocator, 1); // Inter-character gap
    }

    // Stop character
    try result.appendSlice(allocator, &char_patterns[start_stop_idx]);

    return result.toOwnedSlice(allocator);
}

/// Render a Code 39 barcode as PDF content stream operators (filled rectangles).
pub fn render(allocator: Allocator, value: []const u8, x: f32, y: f32, width: f32, height: f32) ![]u8 {
    const bar_widths = try encode(allocator, value);
    defer allocator.free(bar_widths);

    var total_modules: u32 = 0;
    for (bar_widths) |w| {
        total_modules += w;
    }

    const module_width = if (total_modules > 0) width / @as(f32, @floatFromInt(total_modules)) else 0;

    var buf: std.ArrayListUnmanaged(u8) = .{};
    errdefer buf.deinit(allocator);
    const writer = buf.writer(allocator);

    try writer.writeAll("q\n");

    var cur_x = x;
    for (bar_widths, 0..) |w, i| {
        const bar_w = @as(f32, @floatFromInt(w)) * module_width;
        if (i % 2 == 0) {
            // Even indices are bars
            try writer.print("{d:.4} {d:.4} {d:.4} {d:.4} re f\n", .{ cur_x, y, bar_w, height });
        }
        cur_x += bar_w;
    }

    try writer.writeAll("Q\n");

    return buf.toOwnedSlice(allocator);
}

// -- Tests --

test "code39: encode produces output" {
    const allocator = std.testing.allocator;
    const result = try encode(allocator, "HELLO");
    defer allocator.free(result);
    try std.testing.expect(result.len > 0);
}

test "code39: encode wraps with start/stop" {
    const allocator = std.testing.allocator;
    const result = try encode(allocator, "A");
    defer allocator.free(result);
    // Should start with start pattern
    try std.testing.expectEqualSlices(u8, &char_patterns[start_stop_idx], result[0..9]);
}

test "code39: render produces PDF operators" {
    const allocator = std.testing.allocator;
    const result = try render(allocator, "123", 0, 0, 200, 50);
    defer allocator.free(result);
    try std.testing.expect(std.mem.indexOf(u8, result, "q\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, result, "Q\n") != null);
}

test "code39: lowercase converted to uppercase" {
    const allocator = std.testing.allocator;
    const lower = try encode(allocator, "abc");
    defer allocator.free(lower);
    const upper = try encode(allocator, "ABC");
    defer allocator.free(upper);
    try std.testing.expectEqualSlices(u8, upper, lower);
}
